### PR TITLE
Add run scenarios

### LINKED ISSUE #503: Create a list in the documentation of countries with model status (runs, validated, year of validation)
People want to know the country model status e.g. if the optimization works, if the country is validated, when it was validated. We could provide a csv with the [ISO naming](https://www.iso.org/obp/ui/#search) for the countries & populate the table along the way.

| english_short_name | alpha_2_code | alpha_3_code | numeric | optimization_status  | validated            | validation_year | validation_link                  | pypsa_earth_version | GEOMETRY         |
|--------------------|--------------|--------------|---------|----------------------|----------------------|-----------------|----------------------------------|---------------------|------------------|
| str                | str          | str          | int     | bool e.g. True/False | bool e.g. True/False | int e.g. 2022   | str e.g. https://website_link.de | str e.g. "v0.1.0"   | geopandas object |
|                    |              |              |         |                      |                      |                 |                                  |                     |                  |

### PR DESCRIPTION
**Pull request by davide-f**
# Closes #503

## Changes proposed in this Pull Request
This branch aims at creating rules and scripts to iteratively test the workflow on a list of countries.
For each country, the workflow is executed and general statistics are collected.
Intermediate results are also stored.
This can be useful to track the overall status of the workflow for the globe, also collecting statistics that can be useful for validation.

Suggestions are welcome, especially on the statistics to include.

## Checklist

- [x] I tested my contribution locally and it seems to work fine.
- [x] Code and workflow changes are sufficiently documented.
- [x] Newly introduced dependencies are added to `envs/environment.yaml` and `envs/environment.docs.yaml`.
- [x] Changes in configuration options are added in all of `config.default.yaml` and `config.tutorial.yaml`.
- [x] Add a test config or line additions to `test/` (note tests are changing the config.tutorial.yaml)
- [x] Changes in configuration options are also documented in `doc/configtables/*.csv` and line references are adjusted in `doc/configuration.rst` and `doc/tutorial.rst`.
- [x] A note for the release notes `doc/release_notes.rst` is amended in the format of previous release notes, including reference to the requested PR.

### CODE REVIEW

**Comment by ekatef** [comment]
A technical suggestion: would it be possible to keep all the country names for which GADM_ID codes contain non-standard values, that is letters of `GADM_ID` do not correspond to the country code?

It would be very helpful to understand how often fine tuning may be needed when working with administrative data globally. And how much hard-coded could be a solution for non-standard cases :)

**Comment by davide-f** [comment]
> A technical suggestion: would it be possible to keep all the country names for which GADM_ID codes contain non-standard values, that is letters of `GADM_ID` do not correspond to the country code?
> 
> It would be very helpful to understand how often fine tuning may be needed when working with administrative data globally. And how much hard-coded could be a solution for non-standard cases :)

Yeah :) that can be done absolutely, do you have some lines already developed to include that?
I'd just need a line that performs the check including fail-safe check if the datatype does not match.
I was wondering if you may like a test on the time series of renewables. Do you think it is needed?

**Comment by ekatef** [comment]
> Yeah :) that can be done absolutely, do you have some lines already developed to include that?
> I'd just need a line that performs the check including fail-safe check if the datatype does not match.

Super :)
I'd suggest something like https://github.com/ekatef/pypsa-earth/blob/019a49bc95087d3f47fefd6fd8dc155f637cdd91/scripts/build_shapes.py#L145-L175

I have tried here to get use of pandas DataFrame `to_csv()` method with two approaches: 
* one utilising a pre-defined dataframe structure (and thus hopefully safer!),
*  another one writing a filtered-out chunk of the original dataframe (which would be more useful for analysis). 

Agree that we may need an additional safety check to ensure that the workflow will not be stopped due to some I/O errors. But don't have any particular ideas on this (except catching a non-specific exception...) and would be very interested in yours :)

**Comment by ekatef** [comment]
> I was wondering if you may like a test on the time series of renewables. Do you think it is needed?

Absolutely! I'd be more than interested in any kind of validation procedures for renewables. I see here two options:

1) we may try to validate the wind/solar potential calculated for (known) locations of the wind/PV generation against the known generation time series in these places;

2) we can extract climate variables for the locations of the meteorological stations and compare them against the observations and introduce the bias corrections.

Regarding the first point I'm afraid we are somewhat constrained with RES data availability for 2013.

As for the second approach, in my opinion it's a more "correct" way, and there data availability looks much better here. There are great global-scale meteorological datasets like [GHCN](https://www.sciencedirect.com/science/article/pii/S0098300418304230) or [ISD](https://www.ncei.noaa.gov/products/land-based-station/integrated-surface-database). I think it would be very useful to have an opportunity integrate them in our workflow but it may be not trivial at all :)

Probably, as an initial test we could take a couple of dozens locations of meteorological stations (e.g. ones being representative for more or less global picture or just with the highest data quality) and extract wind/solar potential for the buses where these stations belong to? The resulted array could be supplemented with meteorological observations data and used for numerical exercises to obtain a big picture at least in an initial version.

What do you think?

**Comment by davide-f** [comment]
So, I kind of implemented everything on the wish list, at least on a draft way:
- build_shapes: please, check the function `collect_shape_stats` for the statistics that are currently implemented
- build_renewable_profiles: please, check the function `collect_renewable_stats`

They are meant to be simple statistics to check the functionality of the workflow, yet they can be used for validation purposes and easy to expand to include additional features

I think this PR is ready for preliminary review and comparison.
Morever, we should discuss whether to include this to the codebase or not.

@pz-max and @ekatef , will you be available tomorrow for the workflow meeting?

**Comment by ekatef** [comment]
Fantastic! Thank you so much :)
I think both `collect_shape_stats` and `collect_renewable_stats` look great and we'll be able to use their results very well. 

Will be happy to discuss details during the today's workflow meeting

**Comment by pz-max** [comment]
Yes, let's talk today

**Comment by davide-f** [comment]
A run :)
Some work is still needed.
Legend:
- green: network solved
- dark green: at least add_electricity worked, but network was not solved
- yellow: base_network was successful but simplify_network didn't
- orange: build_shapes worked but base_network not
- red: buils_shapes couldn't succeed

img

[Update 3-1-2022]

**Comment by pz-max** [comment]
Nice map @davide-f . Is this figure created by the workflow? (We should regularly check the situation e.g. on weekly basis in near feature). It really shows that pypsa-earth is not yet stable to run everywhere.

**Comment by davide-f** [comment]
> Nice map @davide-f . Is this figure created by the workflow? (We should regularly check the situation e.g. on weekly basis in near feature). It really shows that pypsa-earth is not yet stable to run everywhere.

I'm working on it and we can improve it; it was a nice mapping and testing. I think most of the yellow can be easily fixed.
A lot will be improved when https://github.com/PyPSA/powerplantmatching/pull/96 will be merged and included into conda.
Currently that is running again.

I believe most of the yellow is because of that.
Moreover, most likely some countries will be failing, especially the small ones because of missing data or not enough buses.

Once we have more green for africa, I'll run other countries as well

**Comment by ekatef** [comment]
A nice and very insightful plot!

Would it probably make sense to switch green and dark green to make the color coding more intuitive? (or probably replace gark green with something like gold or yellowgreen?..)

Results are a bit surprising: my a priori feeling was that the model works perfectly in 80% cases and tends to make troubles in really complicated cases only :) Is the yellow probably an effect of #531? And does red mean some regional borders uncertainties?

Agree that the visualisation works perfectly to highlight the modeling status and it's worth to include it into the regular workflow. Apart of this practical meaning, it's just beautiful :)

**Comment by davide-f** [comment]
> A nice and very insightful plot!
> 
> Would it probably make sense to switch green and dark green to make the color coding more intuitive? (or probably replace gark green with something like gold or yellowgreen?..)
> 
> Results are a bit surprising: my a priori feeling was that the model works perfectly in 80% cases and tends to make troubles in really complicated cases only :) Is the yellow probably an effect of #531? And does red mean some regional borders uncertainties?
> 
> Agree that the visualisation works perfectly to highlight the modeling status and it's worth to include it into the regular workflow. Apart of this practical meaning, it's just beautiful :)

Green/dark green currently is not a problem, though I agree.
Before I was using blue.
The major issue now is that we need some little fixing and this mapping is actually something we really needed.
There was a problem with powerplantmatching that I'm investigating and proposed a fix.
The new map should be more realistic and close to the africa paper as well

**Comment by davide-f** [comment]
> Nice map @davide-f . Is this figure created by the workflow? (We should regularly check the situation e.g. on weekly basis in near feature). It really shows that pypsa-earth is not yet stable to run everywhere.

The picture has been created by a notebook.
Maybe it is more suitable for the documentation package.

**Comment by davide-f** [comment]
Image above updated

**Comment by ekatef** [comment]
> Image above updated

Wow 🤩 An impressive progress!

**Comment by davide-f** [comment]
Latest changes are updated above.
The image now looks pretty good. Some bug fixing has been needed in powerplantmatching/atlite and pypsa-earth.
That is the result.

For the purpose of this PR, no more bugfixing is expected, but I'll try run the same code on regions beyond africa.

Major needs to improve the current picture are:
- adjust the clustering when DC buses are included. Probably, the best way would be to do the groupby by ["countries", "carrier"]; that would keep consistency in the carriers by bus without making explode the number of clusters; the effects of sub_networks shall be tested anyway.
- (second priority) Handling empty dataframes where necessary. The workflow failed in some countries because the clustering process occurring on buses without lines seemed to lead errors inside pypsa that may be addressed. An issue as been added in pypsa

To achieve this image, the number of clusters used for the clustering are adjusted by country.
The default used was 5 clusters but that can be relatively high especially for small countries.
A dictionary has been added to track the number of clusters by country.

**Comment by pz-max** [comment]
Update. The picture is becoming more green :1st_place_medal: .. White areas still needs to be tested Screenshot from 2023-01-05 16-17-40 

Legend:
    green: network solved
    dark green: at least add_electricity worked, but network was not solved
    yellow: base_network was successful but simplify_network didn't
    orange: build_shapes worked but base_network not
    red: buils_shapes couldn't succeed

**Comment by davide-f** [comment]
Latest updates.

I think this image cannot be released as of now: too much non-green in Asia.
Maybe the Africa+South America may be released (upper image)

Priorities to be fixed:
1. geometry shape whose draft fix is at https://github.com/pypsa-meets-earth/pypsa-earth/pull/532 . With the revised version and the option drop/set_to_country PK, IN and CH at least should become at least dark green
2. fix cluster network problem at https://github.com/pypsa-meets-earth/pypsa-earth/issues/531 . With this most dark green areas should become clear.

After these fixes, I think the +Asia shape should be ok to be published.
In that case, we can also run Europe and North America

img

**Comment by davide-f** [comment]
This PR will be heavily revised in agreement to the new developments on main.
A force push will update the code and a copy will be stored in my fork

File: scripts/build_test_configs.py [hunk 1]
```diff
@@ -26,6 +26,30 @@ def update(d, u):
     return d
 
 
+def create_test_config(base_path, update_config, output_path):
+    # Input paths
+    fp_baseconfig = Path(Path.cwd(), base_path)
```
Comment (pz-max) [inline scripts/build_test_configs.py]: Docstring

File: Snakefile [hunk 2]
```diff
@@ -835,3 +835,42 @@ rule build_test_configs:
...
     script:
         "scripts/build_test_configs.py"
+
+
+rule run_scenario:
+    input:
+        base_config="config.default.yaml",
+        diff_config="configs/scenarios/config.{scenario_name}.yaml",
+    output:
+        touch("results/{scenario_name}/scenario.done"),
```
Comment (pz-max) [inline Snakefile]: Create a copy of the full config

File: Snakefile [hunk 3]
```diff
@@ -835,3 +835,42 @@ rule build_test_configs:
...
+    threads: 1
+    resources:
+        mem_mb=5000,
+    run:
+        from scripts.build_test_configs import create_test_config
+
+        create_test_config(
```
Comment (pz-max) [inline Snakefile:858]: Short # comment string would be helpful here

File: Snakefile [hunk 4]
```diff
@@ -835,3 +835,42 @@ rule build_test_configs:
         ],
     script:
         "scripts/build_test_configs.py"
+
+
+rule run_scenario:
+    input:
+        base_config="config.default.yaml",
```
Comment (pz-max) [inline Snakefile]: default config instead base_config?

File: Snakefile [hunk 5]
```diff
@@ -835,3 +835,42 @@ rule build_test_configs:
...
+        mem_mb=5000,
+    run:
+        from scripts.build_test_configs import create_test_config
+
+        create_test_config(
+            input.diff_config,
+            {"run": {"name": wildcards.scenario_name}},
+            input.diff_config,
+        )
+        create_test_config(input.base_config, input.diff_config, "config.yaml")
```
Comment (pz-max) [inline Snakefile]: same here, default_config?

File: Snakefile [hunk 6]
```diff
@@ -835,3 +835,42 @@ rule build_test_configs:
...
+    input:
+        expand(
+            "results/{scenario_name}/scenario.done",
+            scenario_name=[
+                c.replace("config.", "").replace(".yaml", "")
+                for c in (
+                    os.listdir("configs/scenarios")
+                    if isdir("configs/scenarios")
+                    else []
+                )
```
Comment (pz-max) [inline Snakefile]: use pathlib instead: https://stackoverflow.com/questions/39909655/listing-of-all-files-in-directory (path.glob seems promising)

**Review by pz-max — CHANGES_REQUESTED** [review]

File: Snakefile (hunk 5 above)
Comment (davide-f) [reply Snakefile]: done

File: Snakefile (hunk 3 above)
Comment (davide-f) [reply Snakefile:858]: done

File: Snakefile (hunk 2 above)
Comment (davide-f) [reply Snakefile]: done

File: Snakefile (hunk 4 above)
Comment (davide-f) [reply Snakefile]: done

File: scripts/build_test_configs.py (hunk 1 above)
Comment (davide-f) [reply scripts/build_test_configs.py]: Done

File: Snakefile (hunk 6 above)
Comment (davide-f) [reply Snakefile]: Done

**Review by pz-max — APPROVED** [review]

**Pull request merged**